### PR TITLE
Implement a semanticdb javac compiler plugin

### PR DESCRIPTION
**Pull request by maxov**
This PR creates a javac compiler plugin in the semanticdb-javac project that generates semanticdb files, much like for semanticdb-scalac. Just like the tests for metac, I also added some expect tests and a diff test with metacp (and its diff with metacp is currently empty, yay!)

Some points of interest:
- In order to properly run the plugin on java compile, I have to [manually set javaHome](https://github.com/scalameta/scalameta/compare/master...maxov:semanticdb-javac#diff-fdc3abdfd754eeb24090dbd90aeec2ceR451). This seems to be related to the fact that sbt doesn't fork the java compiler by default and runs it in the same JVM, which I think means that it doesn't see the plugin on the classpath. However, I was unable to troubleshoot this further because sbt doesn't even spit out errors for javac, perhaps due to https://github.com/sbt/zinc/issues/520.
- I need to do some postprocessing for the javac-metacp [diff test](https://github.com/scalameta/scalameta/compare/master...maxov:semanticdb-javac#diff-1adf10a155e25432e696a2bad27ed9f2R368), because metacp emits names of enum constructor parameters as `$enum$name`, even with the `-parameters` compiler flag

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -155,6 +156,32 @@ lazy val semanticdbScalacPlugin = project
...
   .dependsOn(semanticdbScalacCore)
 
+lazy val semanticdbJavacPlugin = project
+  .in(file("semanticdb/javac"))
+  .settings(
+    moduleName := "semanticdb-javac",
+    description := "Javac compiler plugin that generates SemanticDB on compile",
+    publishableSettings,
+    mergeSettings,
+    fullCrossVersionSettings,
```
Comment (xeno-by) [inline build.sbt]: This project doesn't have to be `CrossVersion.full`. `semanticdbScalac` has to, because it depends on `scala-compiler.jar`, but only because of that.

File: build.sbt [hunk 2]
```diff
@@ -155,6 +156,32 @@ lazy val semanticdbScalacPlugin = project
...
+        }
+        override def transform(node: XmlNode): XmlNodeSeq = node match {
+          case e: Elem if isAbsorbedDependency(node) =>
+            Comment("the dependency that was here has been absorbed via sbt-assembly")
+          case _ => node
+        }
+      }).transform(node).head
+    }
+  )
+  .dependsOn(scalametaJVM, metacp)
```
Comment (xeno-by) [inline build.sbt]: Can you depend on `semanticdbJVM` instead?

Comment (xeno-by) [inline build.sbt]: Also why is the dependency on `metacp` necessary?

File: build.sbt [hunk 3]
```diff
@@ -416,7 +448,21 @@ lazy val semanticdbIntegration = project
         s"-P:semanticdb:synthetics:on"
       )
     },
-    javacOptions += "-parameters"
+    javacSemanticdbDirectory := (target.in(Compile).value / "javac-semanticdb"),
+    javaHome in Compile := {
+      // force javac to fork by setting javaHome, otherwise the compiler errors
+      import scala.sys.process._
+      Some(file("/usr/libexec/java_home -v 1.8".lines_!.head))
```
Comment (xeno-by) [inline build.sbt]: This will fail both AppVeyor and Travis CI builds.

Comment (xeno-by) [inline build.sbt]: @olafurpg Do you have a better idea?

File: build.sbt [hunk 4]
```diff
@@ -394,6 +421,11 @@ lazy val contribJVM = contrib.jvm
 lazy val contribJS = contrib.js
 lazy val contribNative = contrib.native
 
+val javacSemanticdbDirectory =
+  SettingKey[File](
+    "javacSemanticdbDirectory",
+    "location of semanticdb produced by semanticdb-javac")
```
Comment (xeno-by) [inline build.sbt]: Let's hide this in `autoImports` in `project/Build.scala`?

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala [hunk 5]
```diff
@@ -48,6 +48,14 @@ class ExpectSuite extends FunSuite with DiffAssertions {
...
         assertNoDiff(loadObtained, loadExpected)
       }
+      test("javac.expect") {
+        import JavacExpect._
+        assertNoDiff(loadObtained, loadExpected)
+      }
+      test("javac.index") {
+        import JavacIndexExpect._
+        assertNoDiff(loadObtained, loadExpected)
+      }
```
Comment (xeno-by) [inline tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala]: This conflicts with https://github.com/scalameta/scalameta/pull/1682. I'll let @olafurpg decide in which order we want to merge the PRs.

**Comment by olafurpg** [comment]
This is very cool @maxov, I look forward to trying it out! Will try to review tomorrow 😊 I just updated #1682, if there are no more major required changes for that PR I recommend we merge it first and remove the index related diffs here. The removal of the index simplifies SemanticDB generation quite a bit.

File: build.sbt (hunk 2 above)
Comment (maxov) [reply build.sbt]: `SemanticdbIndex` is used

File: build.sbt (hunk 3 above)
Comment (olafurpg) [reply build.sbt]: At the top of my head
```scala
@ sys.props("java.home")
res1: String = "/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre"

@ sys.env("JAVA_HOME")
res2: String = "/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home"
```
system properties are likely more reliable

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/SemanticdbGen.scala [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  def persist(targetRoot: Path): Unit = {
+    val outputFile = targetRoot.resolve("META-INF/semanticdb").resolve(relOutputPath)
+    Files.createDirectories(outputFile.getParent)
+    val outputDocuments = s.TextDocuments(
+      List(
+        s.TextDocument(
+          schema = s.Schema.SEMANTICDB4,
+          uri = relSourcePath.toString,
```
Comment (olafurpg) [inline semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/SemanticdbGen.scala]: Missing `PathIO.toUnix(...)` to work on Windows. Note to self that `RelativePath.toURI` would indeed be nice (https://github.com/scalameta/scalameta/issues/1523)!

**Comment by olafurpg** [comment]
#1682 has now been merged!

**Comment by maxov** [comment]
@olafurpg Rebased on top of your PR. Do you have any ideas how to fix the classpath issue aside from refactoring the resource generators? I'm not sure how to fix the build

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/SemanticdbGen.scala [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  def populate(): Unit = {
+
+    def infoForPackage(pkgSym: String): s.SymbolInformation =
+      s.SymbolInformation(
+        symbol = pkgSym,
+        language = s.Language.JAVA,
+        kind = s.SymbolInformation.Kind.PACKAGE,
+        name = pkgSym.desc.name,
+        accessibility = Some(
```
Comment (olafurpg) [inline semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/SemanticdbGen.scala]: Package should only populate the fields `symbol`, `kind`, `name` 

https://github.com/scalameta/scalameta/blob/master/semanticdb/semanticdb3/semanticdb3.md#root-package

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            elem.populateInfos(infos)
+          }
+          elem.typeParamElements.foreach { elem =>
+            elem.populateInfos(infos)
+          }
+        case elem: ExecutableElement =>
+          elem.paramElements.foreach { elem =>
+            elem.populateInfos(infos)
+          }
+          elem.typeParamElements.foreach { elem =>
```
Comment (olafurpg) [inline semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala]: Nitpick, I would emit type parameters before term parameters.

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    def name: String = elem match {
+      case elem: PackageElement =>
+        if (elem.isUnnamed) "_empty_"
+        else {
+          val fullName = elem.getSimpleName.toString
+          fullName.substring(fullName.lastIndexOf('.') + 1)
+        }
+      case elem => elem.getSimpleName.toString
```
Comment (olafurpg) [inline semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala]: I guess we're guaranteed constructors have `<init>` names?

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+      properties = properties,
+      signature = signature
+    )
+
+    def populateInfos(infos: mutable.ListBuffer[s.SymbolInformation]): s.SymbolInformation = {
+      val myInfo = info
+      infos += myInfo
+      elem match {
+        case elem: TypeElement =>
+          elem.enclosedElements.foreach { elem =>
```
Comment (olafurpg) [inline semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala]: What happens to static initializers here? I think metacp excludes them but there are no tests for it. Static initializers have `<cinit>` name and no signature.

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala (hunk 9 above)
Comment (maxov) [reply semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala]: Yes. From the javac api:
```
For a constructor, the name {@code "<init>"} is
returned, for a static initializer, the name {@code "<clinit>"}
is returned, and for an anonymous class or instance
initializer, an empty name is returned.
```

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/SemanticdbGen.scala (hunk 7 above)
Comment (maxov) [reply semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/SemanticdbGen.scala]: Metacp also populates the language and accessibility fields, so I added this and it expanded the diff with metacp.

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala (hunk 10 above)
Comment (maxov) [reply semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala]: Javac also excludes them (they are not listed in "enclosed elements'), I can try to figure out how to avoid that but it doesn't look trivial.

File: semanticdb/semanticdb3/guide.md [hunk 11]
```diff
@@ -357,6 +358,22 @@ addCompilerPlugin("org.scalameta" % "semanticdb-scalac" % "4.0.0-M5" cross Cross
 scalacOptions += "-Yrangepos"
 ```
 
+### Javac compiler plugin
+
+The `semanticdb-javac` compiler plugin acts harvests and dumps SemanticDB information after the analyze phase of the Java compiler.
+
+To use it, add it to your build's compile classpath and add the following javac option:
+
+```"-Xplugin:semanticdb <target-dir> --sourceroot <source-root>"```
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply semanticdb/semanticdb3/guide.md]: Might be worth to stress that the quotes are necessary, this should be a single string argument. I originally thought the quotes were there by mistake

File: semanticdb/semanticdb3/guide.md [hunk 12]
```diff
@@ -357,6 +358,22 @@ addCompilerPlugin("org.scalameta" % "semanticdb-scalac" % "4.0.0-M5" cross Cross
 scalacOptions += "-Yrangepos"
 ```
 
+### Javac compiler plugin
+
+The `semanticdb-javac` compiler plugin acts harvests and dumps SemanticDB information after the analyze phase of the Java compiler.
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply semanticdb/semanticdb3/guide.md]: Might be good to mention that it only produces the "Symbols" section at the moment. I opened #1692 to track "Occurrences"

File: semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala (hunk 9 above)
Comment (olafurpg) [reply semanticdb/javac/src/main/scala/scala/meta/internal/semanticdb/javac/semantics/Elements.scala]: 👍

File: semanticdb/semanticdb3/guide.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (maxov) [reply semanticdb/semanticdb3/guide.md]: It doesn't appear that compiler options are part of the supported javac API, so I cannot get the output directory for classfiles.
